### PR TITLE
net: Replace enum CConnMan::NumConnections with enum class ConnectionDirection

### PR DESCRIPTION
**Pull request by luke-jr**
Refactor split out of #17167

### CODE REVIEW

File: src/netbase.h [hunk 1]
```diff
@@ -25,6 +25,22 @@ static const int DEFAULT_CONNECT_TIMEOUT = 5000;
...
 static const int DEFAULT_NAME_LOOKUP = true;
 
+enum class ConnectionDirection {
+        None = 0,
+        In = (1U << 0),
+        Out = (1U << 1),
+        Both = (In | Out),
+};
+static inline ConnectionDirection& operator|=(ConnectionDirection& a, ConnectionDirection b) {
+    using underlying = typename std::underlying_type<ConnectionDirection>::type;
```
Comment (promag) [inline src/netbase.h:35]: Add `#include <type_traits>`.

File: src/netbase.h [hunk 2]
```diff
@@ -25,6 +25,22 @@ static const int DEFAULT_CONNECT_TIMEOUT = 5000;
...
+    a = ConnectionDirection(underlying(a) | underlying(b));
```
Comment (promag) [inline src/netbase.h:36]: nit, `return a = ...`?

Comment (luke-jr) [reply src/netbase.h:36]: I think it's more readable the current way

File: src/netbase.h (hunk 1 above)
Comment (luke-jr) [reply src/netbase.h:35]: k

File: src/netbase.h [hunk 3]
```diff
@@ -25,6 +26,22 @@ static const int DEFAULT_CONNECT_TIMEOUT = 5000;
 //! -dns default
 static const int DEFAULT_NAME_LOOKUP = true;
 
+enum class ConnectionDirection {
+        None = 0,
+        In = (1U << 0),
+        Out = (1U << 1),
+        Both = (In | Out),
```
Comment (MarcoFalke) [inline src/netbase.h:33]: Any reason the new code is not clang formatted?

**Review by promag — COMMENTED** [review]
Code review ACK 8727019ec307badb3bca363cc8bc832aaea4c058, just needs to format correctly apparently.

File: src/netbase.h (hunk 3 above)
Comment (luke-jr) [reply src/netbase.h:33]: I don't see any documented way to do that (`contrib/devtools/clang-format-diff.py` wants to make a bunch of unrelated changes).

Fixed indentation. Anything else?

File: src/netbase.h [hunk 4]
```diff
@@ -25,6 +26,22 @@ static const int DEFAULT_CONNECT_TIMEOUT = 5000;
...
+};
+static inline ConnectionDirection& operator|=(ConnectionDirection& a, ConnectionDirection b) {
+    using underlying = typename std::underlying_type<ConnectionDirection>::type;
+    a = ConnectionDirection(underlying(a) | underlying(b));
+    return a;
+}
+static inline bool operator&(ConnectionDirection a, ConnectionDirection b) {
+    using underlying = typename std::underlying_type<ConnectionDirection>::type;
+    return (underlying(a) & underlying(b));
+}
```
Comment (MarcoFalke) [inline src/netbase.h:43]: Seems a bit odd to add unused code, but no strong opinion.

**Review by maflcko — APPROVED** [review]
ACK

**Comment by practicalswift** [comment]
ACK 40bc8d725f6542697a0580df0087b230959a3db7: patch looks correct

**Comment by practicalswift** [comment]
cr ACK c77de622dd8ef458f73b1a01a34629a7c4f49358: patch looks correct & `enum class` is strictly better

File: src/netbase.h (hunk 4 above)
Comment (jonatack) [reply src/netbase.h:46]: > Seems a bit odd to add unused code, but no strong opinion.

If I understand the comment correctly, this `operator&` code is used in `src/net.cpp:2736`
```cpp
        if (flags & (pnode->IsInboundConn() ? ConnectionDirection::In : ConnectionDirection::Out)) {
```

**Pull request merged**